include/zephyr/drivers/charger.h

```diff
@@ -0,0 +1,196 @@
+	CHARGER_STATUS_FULL,
+};
+
+/**
+ * @typedef charger_property_t
+ * @brief A charger property's identifier
+ *
+ * See charger_props for a list of identifiers
+ */
+typedef uint16_t charger_property_t;
+
+/**
+ * @brief Representation of a charger property
+ *
+ * The property has a property type that identifies the property to get,
+ * an error associated with fetching the property, and the fetched value
+ * of the property.
+ */
+struct charger_get_property {
+	/** Battery charger property to get */
+	charger_property_t property_type;
+	/** Negative error status set by callee e.g. -ENOTSUP for an unsupported property */
+	int err;
+	/** Property field for getting */
+	union {
+		enum charger_online online;
+		bool present;
+		enum charger_status status;
+	} value;
+};
+
+/**
+ * @typedef charger_get_properties_t
+ * @brief Callback API for getting multiple charger properties.
+ *
+ * See charger_get_property() for argument description
+ */
+typedef int (*charger_get_properties_t)(const struct device *dev,
+					struct charger_get_property *props, size_t props_len);
+
+/**
+ * @typedef charger_get_property_t
+ * @brief Callback API for getting a charger property.
+ *
+ * See charger_get_property() for argument description
+ */
+typedef int (*charger_get_property_t)(const struct device *dev, struct charger_get_property *prop);
+
+/**
+ * @brief Charging device API
+ *
+ * Caching is entirely on the onus of the client
+ */
+__subsystem struct charger_driver_api {
+	charger_get_property_t get_property;
+};
+
+/**
+ * @brief Fetch a battery charger property
+ *
+ * @param dev Pointer to the battery charger device
+ * @param prop pointer to charger_get_property struct where the property struct
+ * field is set by the caller to determine what property is read from the
+ * charger device into the charger_get_property struct's value field.
+ *
+ * @retval 0 if successful
+ * @retval < 0 if getting property failed
+ */
+__syscall int charger_get_prop(const struct device *dev, struct charger_get_property *prop);
+
+static inline int z_impl_charger_get_prop(const struct device *dev,
+					  struct charger_get_property *prop)
+{
+	const struct charger_driver_api *api = (const struct charger_driver_api *)dev->api;
+
+	return api->get_property(dev, prop);
+}
+
+/**
+ * @brief Fetch multiple battery charger properties
+ *
+ * @param dev Pointer to the battery charger device
+ * @param props pointer to array of charger_get_property struct where the property struct
+ * field is set by the caller to determine what property is read from the
+ * charger device into the charger_get_property struct's value field. The array is not
+ * permuted and elements can be directly accessed as loaded.
+ * @param props_len number of properties in props array
+ *
+ * @retval 0 if successful
+ * @retval < 0 if getting all properties failed
+ * @retval > 0 if some properties failed where return=number of failing properties.
```

Comment:
note: with this mechanism you'll never know the original error condition, I'd suggest either:

1. fail fast, and propagate the first error condition (this would actually make the struct redundant, as there'd be no need to store each err)
2. return the number of obtained properties, not sure if useful for clients of the API, though.

Reply:
I'll go with option #2